### ISSUE TITLE
[core] Identifiers Too Similiar

### ISSUE THREAD

**Issue by nathanila** [pmd/pmd#2094]
**Affects PMD Version:** 

6.18.0

**Rule:**

New Rule: IdentifiersTooSimiliar

**Description:**

This new rule will flag a problem if two identifiers are different by only N characters where N is configurable.  The default value for N should be 1.

This rule is applicable for Java and probably all languages.

Let's say I have two local variables named "col" and "co1" in a single method.  These identifiers are almost identical except for 1 character difference.  When reading the code, it is easy to miss the difference and interpret "col" as "co1" or "co1" as "col".

A more common situation is "columns" and "column".  The only difference is an "s".  Again, it is easy to miss the difference and misinterpret the variable.

Some times the difference is in the middle of the identifier.  Human brains are very likely to miss this.  For example, these two identifiers are only different by the "s" in the middle: colsToIndex and colToIndex.

The problem also occurs with fields vs local variables, fields vs fields, methods vs methods and classes vs classes.  I am not sure if there is a problem with variables vs methods, variables vs classes or methods vs classes.  These later potential conflicts will need some thought.

PMD should flag these situations as problems so that the programmer will pick a different name for one or both identifiers and improve code readability.

**Code Sample demonstrating the issue:**

```
{
   List<Column> columns;
   Column column;

   ...

   // Let's assume that Column.size() exists.
   // The compiler won't flag a problem but columns.size() should have been called.
   // The programmer missed the problem because the two identifiers are too similar.
   for (int i = 0; i < column.size(); i++)
   {
      column = columns.get(i);
      // ...
   }
}
```

**Running PMD through:**

Not applicable

**Comment by blerner** [comment]
Can we refine this suggestion to flag overly-similar names _when they have the same type_?  It seems somewhat reasonable to write `for (Column col : cols) { ... }` and the type-checker intrinsically prevents conflating one with the other (unless you do something that mucks about with `toString`, numbers or other implicit conversions).  I wouldn't want to unintentionally encourage my students to write `for (Column c : cols)` just to make it appear to be three characters different...!

Depending on how much type information is available to this rule, I'd be fine complaining when one name has the same type as or a subtype of the other, but I don't know if the type hierarchy is available here.

**Comment by nathanila** [comment]
`col` vs `cols` is exactly the kind of problem I am trying to flag.  Yes, the types are different but sometimes they share the same method names.  (See my example in the original post.)  This leads to subtle bugs.

As for people using short names, the PMD rules [ShortVariable](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#shortvariable), [ShortMethodName](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#shortmethodname) and [ShortClass](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#shortclassname) will discourage using `c` or `co`.

**Comment by adangel** [comment]
We could make the rule configurable, so that the user can decide, whether the types should be considered or not.

By the way: To which category would this rule belong to? Codestyle?

**Comment by nathanila** [comment]
> We could make the rule configurable, so that the user can decide, whether the types should be considered or not.

That's a good idea.

Code Style has a lot of rules pertaining to variables and their naming.

**Comment by Fernal73** [comment]
Are there any cases where you'd actually wish to use or couldn't avoid similar names?

"Character Name

L 0011
L_0111
L_1001
L_1100
L_0101
L_1010
L_1212"

http://www.oualline.com/style/c03.html

I can also think of variables like col1,col1, col3 or row1,row2,row2

But these could be changed to colOne,colTwo, colThree

You definitely don't want:
coli, colii, coliii, coliv,colv...

**Comment by nathanila** [comment]
> Are there any cases where you'd actually wish to use or couldn't avoid similar names?

I use `i`, `j` and sometimes `k` for loop iterators.  This new rule would flag a problem with these variables.

> I can also think of variables like col1, col1, col3 or row1, row2, row2

Case in point.  You had two typos in the variable names above.  The names are too similar that you did not catch the mistake.

Moving on... I had not thought about this case.  I suppose in the worst case the person could add `//NOPMD` to the line or add `@SuppressWarnings("PMD.IdentifiersTooSimiliar")` to the method if they thought their particular use case warrants an exception.  Perhaps, the rule would have an `allowed` configuration option where the user can specify a regex expression of identifier names to allow and not flag as a problem.

**Comment by Fernal73** [comment]
It wasn't a typo. It was auto-correct.

**Comment by nathanila** [comment]
I've heard of people doing analysis on open source code.  Is there an easy way to analyze open source code?  I would want to see how often would the rule trigger and what the identifiers involved would be.  We could then see what percent of methods would be flagged by this rule and look at the identifier names.

**Comment by Fernal73** [comment]
The examples I quoted:
The first is sufficiently rare. That can be suppressed.
The second is common. It ought to be flagged by this rule. Would it be a pain-point for programmers? Like would they go, 'WTH?' and disable the rule?

**Comment by Fernal73** [comment]
Your suggestion is a project by itself and could make for a great research paper.

**Comment by nathanila** [comment]
It is hard for me to say how programmers will react to the rule.  I have little experience in this area.

As for using an array instead, perhaps that is another performance rule.  If someone creates an array in a method and it is only used locally (and not in a lambda), then the performance rule should flag it.

**Comment by Fernal73** [comment]
I didn't realize that it could become a rule. It's simply that if a rule
like yours exists and I'd have to rename all my numerical-suffixed
variables, an array would be a substitute especially since it's obvious
that I do know how many variables would be required. It's a circumvention.
It wouldn't matter if reflection was being used to create variables on the
fly dynamically. Just a few idle thoughts.

**Comment by tweimer** [comment]
> It is hard for me to say how programmers will react to the rule. I have little experience in this area.

I would definately disable that rule.  See example by @blerner and your example about loop indices.

**Comment by Fernal73** [comment]
I agree with @blerner.
I was more concerned with local variables having the same name or similar names as instance variables. Same names can be mitigated by using 'this' pointer. It's the similar names that will be a concern. Or will this rule check only for local variables?

**Comment by Fernal73** [comment]
Additionally , at what level of granularity do you wish an error to be flagged?
Should we disallow similar names on the same line? Within the same code block (within curly braces)? Or within the same method? Could we adjust the rule to weigh them accordingly? Error in the first case, warning in the next and informational in the last?
